src/v/kafka/server/server.cc

```diff
@@ -129,14 +129,38 @@ server::server(
   , _gssapi_principal_mapper(
       config::shard_local_cfg().sasl_kerberos_principal_mapping.bind())
   , _krb_configurator(config::shard_local_cfg().sasl_kerberos_config.bind())
+  , _memory_fetch_sem(
+      static_cast<size_t>(
+        cfg->local().max_service_memory_per_core
+        * config::shard_local_cfg().kafka_memory_share_for_fetch()),
+      "kafka/server-mem-fetch")
   , _thread_worker(tw) {
+    vlog(klog.debug, "memory_fetch: {}", _memory_fetch_sem.available_units());
     if (qdc_config) {
         _qdc_mon.emplace(*qdc_config);
     }
+    setup_metrics();
     _probe.setup_metrics();
     _probe.setup_public_metrics();
 }
 
+void server::setup_metrics() {
+    namespace sm = ss::metrics;
+    if (config::shard_local_cfg().disable_metrics()) {
+        return;
+    }
+
+    _metrics.add_group(
+      prometheus_sanitize::metrics_name(cfg.name),
+      {
+        sm::make_total_bytes(
+          "fetch_avail_mem_bytes",
+          [this] { return _memory_fetch_sem.current(); },
+          sm::description(ssx::sformat(
+            "{}: Memory available for fetch request processing", cfg.name))),
```

Comment:
This metric looks a bit lonely, maybe we could also track:
- low watermark of available units, so we can tell if a system has been limited by this semaphore
- the main memory semaphore as well, unless that's already reported somewhere else.

Since these are low-cardinality per-shard metrics, they could be registered as public metrics rather than internal ones.

Reply:
The kafka mem semaphore is reported by `net::server`. It is a non-public metric, that's why this one is made non-public too. Would you suggest making them both public? That would also affect metrics for internal RPC memsemaphore.

+1 for the LWM reporting.